### PR TITLE
[FIX] queue_job: prevent issues with native test TestOverrides:test_creates()

### PR DESCRIPTION
**Pull request by Kimkhoi3010**
### Issue: 
The native test [TestOverrides:test_creates()](https://github.com/odoo/odoo/blob/3c63c517395b758b69c70a99f157501d3da0953e/odoo/addons/base/tests/test_overrides.py#L13 ) fails when queue_job is installed due to a security check that prevents the creation of queue.job records via RPC (see [queue_job.py#L234](https://github.com/OCA/queue/blob/18.0/queue_job/models/queue_job.py#L234)). This results in an exception:
```
ERROR: TestOverrides.test_creates
Traceback (most recent call last):
  File "/home/odoo/src/odoo/odoo/addons/base/tests/test_overrides.py", line 19, in test_creates
    model_env.create([]), model_env.browse(),
    ^^^^^^^^^^^^^^^^^^^^
  File "<decorator-gen-160>", line 2, in create
  File "/home/odoo/src/odoo/odoo/api.py", line 480, in _model_create_multi
    return create(self, arg)
           ^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/user/queue_job_cron_jobrunner/models/queue_job.py", line 147, in create
    records = super().create(vals_list)
              ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<decorator-gen-155>", line 2, in create
  File "/home/odoo/src/odoo/odoo/api.py", line 480, in _model_create_multi
    return create(self, arg)
           ^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/user/queue_job/models/queue_job.py", line 237, in create
    raise exceptions.AccessError(
odoo.exceptions.AccessError: Queue jobs must be created by calling 'with_delay()'.
```

### Solution: 

- This PR modifies the create method in the queue_job model for that specific case, and only that one (test mode + empty vals_list).

### CODE REVIEW

File: queue_job/models/queue_job.py [hunk 1]
```diff
@@ -231,6 +231,12 @@ def _compute_graph_jobs_count(self):
 
     @api.model_create_multi
     def create(self, vals_list):
+        if not vals_list:
+            return super(
+            QueueJob,
+            self.with_context(mail_create_nolog=True, mail_create_nosubscribe=True),
+        ).create(vals_list)
+
         if self.env.context.get("_job_edit_sentinel") is not self.EDIT_SENTINEL:
```
Comment (nilshamerlinck) [inline queue_job/models/queue_job.py:239]: so the new block above is the same as adding a new predicate `if vals_list and` in this line no?

**Comment by nilshamerlinck** [comment]
- could you please rebase this one on 18.0 branch? (as this is a fix on `queue_job`)
- `[UPD] - No need to check when creating with empty values]`: please improve commit name by having a look at the commits naming conventions of OCA: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#commit-message
- please also come up with a PR description that explains why you are fixing that 
- more importantly, your fix here means that the security check is bypassed: did you try to "create a queue.job record "raw" from RPC" (call `queue.job:create()`)?

File: queue_job/models/queue_job.py [hunk 2]
```diff
@@ -231,6 +232,12 @@ def _compute_graph_jobs_count(self):
 
     @api.model_create_multi
     def create(self, vals_list):
+        test_mode = getattr(threading.current_thread(), 'testing', False) or self.env.registry.in_test_mode()
```
Comment (nilshamerlinck) [inline queue_job/models/queue_job.py]: why both? in which case `getattr(threading.current_thread(), 'testing', False) != self.env.registry.in_test_mode()`?

File: queue_job/models/queue_job.py [hunk 3]
```diff
@@ -231,6 +232,12 @@ def _compute_graph_jobs_count(self):
...
+        if not vals_list and test_mode:
+            return super(
+            QueueJob,
+            self.with_context(mail_create_nolog=True, mail_create_nosubscribe=True),
+        ).create(vals_list)
```
Comment (nilshamerlinck) [inline queue_job/models/queue_job.py]: why not as a predicate in next line?
as per DRY principles

**Review by nilshamerlinck — CHANGES_REQUESTED** [review]
really need to rebase, there should only be one FIX commit

**Comment by nilshamerlinck** [comment]
your PR description is in the good direction, but very verbose
I think you can simplify it:

- native test [`TestOverrides:test_creates()`](https://github.com/odoo/odoo/blob/18.0/odoo/addons/base/tests/test_overrides.py#L13
) fails with `queue_job` installed because of the [security check that prevents creation of `queue.job` records via rpc (add link to code here):

{include exception here}

- this PR bypasses the check in test context (+ more explanation for `vals_list`)

**Comment by nilshamerlinck** [comment]
PR description:
- can you please remove the screenshot and copy-paste the exception only?

> This PR bypasses the security check in test contexts (e.g., in_test_mode or threading.testing). Additionally, it handles empty vals_list gracefully by returning model.browse() during tests, ensuring compatibility with test cases that validate create([]).

not clear
- why is (1) not enough?
- which tests cases are you refering to for (2)?
- `in_test_mode`: I believe you removed that bit from code already so not sure why you mention it

File: queue_job/models/queue_job.py [hunk 4]
```diff
@@ -231,6 +232,9 @@ def _compute_graph_jobs_count(self):
 
     @api.model_create_multi
     def create(self, vals_list):
+        test_mode = getattr(threading.current_thread(), 'testing', False)
+        if not vals_list and test_mode:
+            return super().create(vals_list)
```
Comment (nilshamerlinck) [inline queue_job/models/queue_job.py]: so we loose `with_context(mail_create_nolog=True, mail_create_nosubscribe=True)`?
needed?

**Comment by nilshamerlinck** [comment]
one more thing: `[FIX] queue_job: handle test case for empty create vals_list`
does it really summarize the fix properly?
I would say:
`[FIX] queue_job: prevent issues with native test TestOverrides:test_creates()`

**Comment by nilshamerlinck** [comment]
why you don't factorize it as follow?

```
        if not test_mode and self.env.context.get("_job_edit_sentinel") is not self.EDIT_SENTINEL:
            # Prevent to create a queue.job record "raw" from RPC.
            # ``with_delay()`` must be used.
            raise exceptions.AccessError(
                _("Queue jobs must be created by calling 'with_delay()'.")
            )
        return super(
            QueueJob,
            self.with_context(mail_create_nolog=True, mail_create_nosubscribe=True),
        ).create(vals_list)
```

also, you didn't explain why the condition on `vals_list`, isn't the check on `test_mode` enough?

**Comment by Kimkhoi3010** [comment]
> why you don't factorize it as follow?
> 
> ```
>         if not test_mode and self.env.context.get("_job_edit_sentinel") is not self.EDIT_SENTINEL:
>             # Prevent to create a queue.job record "raw" from RPC.
>             # ``with_delay()`` must be used.
>             raise exceptions.AccessError(
>                 _("Queue jobs must be created by calling 'with_delay()'.")
>             )
>         return super(
>             QueueJob,
>             self.with_context(mail_create_nolog=True, mail_create_nosubscribe=True),
>         ).create(vals_list)
> ```
> 
> also, you didn't explain why the condition on `vals_list`, isn't the check on `test_mode` enough?

I tried the way you mentioned and got an error at [test_queue_job_protected_write.py](https://github.com/OCA/queue/blob/eb5825327c005bca970641bc557fa54c3aeac9d8/queue_job/tests/test_queue_job_protected_write.py#L9)
This error will appear:
```
ERROR	server	bad query: b'INSERT INTO "queue_job" ("method_name", "model_name", "uuid") VALUES (\'write\', \'res.partner\', \'test\') RETURNING "id"'
ERROR: null value in column "state" of relation "queue_job" violates not-null constraint
DETAIL:  Failing row contains (1, null, null, null, null, null, null, null, test, null, null, res.partner, write, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null).
09/12/2024 10:52:44	ERROR	server	ERROR: TestJobWriteProtected.test_create_error
Traceback (most recent call last):
  File "/home/odoo/src/user/queue_job/tests/test_queue_job_protected_write.py", line 11, in test_create_error
    self.env["queue.job"].create(
  File "<decorator-gen-157>", line 2, in create
  File "/home/odoo/src/odoo/odoo/api.py", line 479, in _model_create_multi
    return create(self, [arg])
           ^^^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/user/queue_job/models/queue_job.py", line 248, in create
    ).create(vals_list)
      ^^^^^^^^^^^^^^^^^
  File "<decorator-gen-47>", line 2, in create
  File "/home/odoo/src/odoo/odoo/api.py", line 480, in _model_create_multi
    return create(self, arg)
           ^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/odoo/addons/mail/models/mail_thread.py", line 267, in create
    threads = super(MailThread, self).create(vals_list)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<decorator-gen-0>", line 2, in create
  File "/home/odoo/src/odoo/odoo/api.py", line 480, in _model_create_multi
    return create(self, arg)
           ^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/odoo/odoo/models.py", line 4957, in create
    records = self._create(data_list)
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/odoo/odoo/models.py", line 5141, in _create
    cr.execute(SQL(
  File "/home/odoo/src/odoo/odoo/sql_db.py", line 354, in execute
    res = self._obj.execute(query, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
psycopg2.errors.NotNullViolation: null value in column "state" of relation "queue_job" violates not-null constraint
DETAIL:  Failing row contains (1, null, null, null, null, null, null, null, test, null, null, res.partner, write, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null).
```
Therefore, to ensure the logic of the test cases, I chose to rely on val_list in addition to relying on test_mode

**Comment by nilshamerlinck** [comment]
I understand but given that we have 2 blocks that do the same thing:

(1)

```
            self = self.with_context(
                mail_create_nolog=True, mail_create_nosubscribe=True
            )
            res = super().create(vals_list)
            return res
```

(2)
   
```
        return super(
            QueueJob,
            self.with_context(mail_create_nolog=True, mail_create_nosubscribe=True),
        ).create(vals_list)
``` 

don't you think that you can factorize things a bit?

**Comment by nilshamerlinck** [comment]
I had a second look at this problem, I think we should simplify the fix a bit

Solution:
- This PR modifies the create method in the queue_job model for that specific case, and only that one (test mode + empty vals_list).

Can you try with just that addition?

```
    @api.model_create_multi
    def create(self, vals_list):
        test_mode = getattr(threading.current_thread(), "testing", False)
        if test_mode and not vals_list:
            return self
```

**Review by nilshamerlinck — APPROVED** [review]
does it work as expected? if so, ok for PR on oca side, but please update description (see my last comment)
